src/garage/tf/optimizers/conjugate_gradient_optimizer.py

```diff
@@ -1,27 +1,118 @@
+computes the optimal step size that will satisfy the KL divergence constraint.
+Finally, it performs a backtracking line search to optimize the objective.
+
+"""
+import abc
+
 from dowel import logger
 import numpy as np
 import tensorflow as tf
 
-from garage.core import Serializable
-from garage.misc import krylov
 from garage.tf.misc import tensor_utils
 from garage.tf.optimizers.utils import LazyDict, sliced_fun
 
 
-class PerlmutterHvp:
+class HessianVectorProduct(abc.ABC):
+    """Base class for computing Hessian-vector product.
+
+    Args:
+        num_slices (int): Hessian-vector product function's inputs will be
+            divided into num_slices and then averaged together to improve
+            performance.
+
+    """
+
     def __init__(self, num_slices=1):
-        self.target = None
-        self.reg_coeff = None
-        self.opt_fun = None
+        self._target = None
+        self._reg_coeff = None
+        self._hvp_fun = None
         self._num_slices = num_slices
 
-    def update_opt(self, f, target, inputs, reg_coeff, name=None):
-        self.target = target
-        self.reg_coeff = reg_coeff
+    @abc.abstractmethod
+    def update_hvp(self, f, target, inputs, reg_coeff, name=None):
+        """Build the symbolic graph to compute the Hessian-vector product.
+
+        Args:
+            f (tf.Tensor): The function whose Hessian needs to be computed.
+            target (garage.tf.policies.Policy): A parameterized object to
+                optimize over.
+            inputs (tuple[tf.Tensor]): The inputs for function f.
+            reg_coeff (float): A small value so that A -> A + reg*I.
+            name (str): Name to be used in tf.name_scope.
+
+        """
+
+    def build_eval(self, inputs):
+        """Build the evaluation function.  # noqa: D202, E501 # https://github.com/PyCQA/pydocstyle/pull/395.
+
+        Args:
+            inputs (tuple[numpy.ndarray]): Function f will be evaluated on
+                these inputs.
+
+        Returns:
+            function: It can be called to get the final result.
+
+        """
+
+        def _eval(v):
+            """The evaluation function.
+
+            Args:
+                v (numpy.ndarray): The vector to be multiplied with Hessian.
+
+            Returns:
+                numpy.ndarray: The product of Hessian of function f and v.
+
+            """
+            xs = tuple(self._target.flat_to_params(v, trainable=True))
+            ret = sliced_fun(self._hvp_fun['f_hx_plain'], self._num_slices)(
+                inputs, xs) + self._reg_coeff * v
+            return ret
+
+        return _eval
+
+    def __getstate__(self):
+        """Object.__getstate__.
+
+        Returns:
+            dict: the state to be pickled for the instance.
+
+        """
+        new_dict = self.__dict__.copy()
+        del new_dict['_hvp_fun']
+        return new_dict
+
+
+class PerlmutterHvp(HessianVectorProduct):
```

Comment:
This is probably pedantic. But I think there is a typo in the class name here and the tests. The three spellings in the function name, the name in the citation and the name of the test class are all different. I think It should be Pearlmutter - as in the citation.

Reply:
Ah..Thanks for pointing out. Fixed it.